src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -14,71 +14,72 @@ namespace System.Buffers
     public readonly partial struct ReadOnlySequence<T>
     {
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal bool TryGetBuffer(in SequencePosition start, in SequencePosition end, out ReadOnlyMemory<T> data, out SequencePosition next)
+        internal bool TryGetBuffer(in SequencePosition position, out ReadOnlyMemory<T> memory, out SequencePosition next)
         {
-            SequenceType type;
-            int startIndex = 0;
-            int endIndex = 0;
+            object positionObject = position.GetObject();
             next = default;
-            object startObject = start.GetObject();
-            if (startObject != null)
-            {
-                GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out type, out startIndex, out endIndex);
-            }
-            else
+
+            if (positionObject == null)
             {
-                type = SequenceType.Empty;
+                memory = default;
+                return false;
             }
 
-            int length = endIndex - startIndex;
-            object endObject = end.GetObject();
-
-            if (type != SequenceType.MultiSegment && type != SequenceType.Empty && startObject != endObject)
```

Comment:
Why is this not needed anymore?

Reply:
I see, you inlined them, why?

Reply:
To make the MultiSegment case faster. We are already doing `type == X` checks. Doing this check up front adds a redundant comparison.

Reply:
As far as I can tell, I made no changes that would affect the semantics of the method. At least that was my intention.

Reply:
Can we put 
```
positionObject != endObject)
   ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
```

into else branch of  `if (type == SequenceType.MultiSegment)` so we call it once instead of three times?

Reply:
I considered that. Let me verify that it doesn't have any negative performance side effects.

Reply:
No significant change one way or another (I will make the change to avoid code duplication):
![image](https://user-images.githubusercontent.com/6527137/38644128-5df08cbe-3d94-11e8-92a2-c0616a26cd93.png)

Reply:
This table shows worse regression than the first one you've posted :)

Reply:
That is diff only for the if/else change you suggested (I wanted to show the isolated results). The original performance improvement still holds

Reply:
Nice, good job then.